File: ch07/VbServerControl1/Listing07-34.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Text
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls

<DefaultProperty("Text")>
<ToolboxData("<{0}:Listing0734 runat=server></{0}:Listing0734>")>
Public Class Listing0734
    Inherits System.Web.UI.WebControls.WebControl

    <Browsable(False)> Public Property TemplateMessage() As Message

    <PersistenceMode(PersistenceMode.InnerProperty),
        TemplateContainer(GetType(Message))>
    Public Property MessageTemplate() As ITemplate

    <Bindable(True), DefaultValue("")>
    Public Property Name() As String

    <Bindable(True), DefaultValue("")>
    Public Property Text() As String

    Public Overrides Sub DataBind()
        EnsureChildControls()
        ChildControlsCreated = True

        MyBase.DataBind()
    End Sub

    Protected Overrides Sub CreateChildControls()
        Me.Controls.Clear()
        Me.TemplateMessage = New Message() With {.Name = Name, .Text = Text}

        If Me.MessageTemplate Is Nothing Then
            Me.MessageTemplate = New DefaultMessageTemplate()
        End If

        Me.MessageTemplate.InstantiateIn(Me.TemplateMessage)
        Controls.Add(Me.TemplateMessage)
    End Sub

    Protected Overrides Sub RenderContents(
        ByVal writer As System.Web.UI.HtmlTextWriter)

        EnsureChildControls()
        ChildControlsCreated = True

        MyBase.RenderContents(writer)
    End Sub
End Class

File: ch07/CSharpServerControl1/Listing07-34.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace CSharpServerControl1
{
    [DefaultProperty("Text")]
    [ToolboxData("<{0}:Listing0734 runat=server></{0}:Listing0734>")]
    public class Listing0734 : WebControl
    {
        [Browsable(false)]
        public Message TemplateMessage { get; internal set; }

        [PersistenceMode(PersistenceMode.InnerProperty)]
        [TemplateContainer(typeof(Message))]
        public virtual ITemplate MessageTemplate { get; set; }

        [Bindable(true)]
        [DefaultValue("")]
        public string Name { get; set; }

        [Bindable(true)]
        [DefaultValue("")]
        public string Text { get; set; }

        public override void DataBind()
        {
            EnsureChildControls();
            ChildControlsCreated = true;

            base.DataBind();
        }

        protected override void CreateChildControls()
        {
            this.Controls.Clear();
            this.TemplateMessage = new Message() { Name = Name, Text = Text };

            if (this.MessageTemplate == null)
            {
                this.MessageTemplate = new DefaultMessageTemplate();
            }

            this.MessageTemplate.InstantiateIn(this.TemplateMessage);
            Controls.Add(this.TemplateMessage);
        }

        protected override void RenderContents(HtmlTextWriter writer)
        {
            EnsureChildControls();
            ChildControlsCreated = true;

            base.RenderContents(writer);
        }
    }
}